### PR TITLE
 Adopt AMQ::Settings from amq-protocol gem

### PR DESCRIPTION
**Pull request by Tensho**
Addresses ruby-amqp/amqp/issues/236

Actually `amqp` doesn't negotiate `chanel_max` and doesn't handle `connection_timeout`, so I added only `heartbeat` option after URI parsing.

Cross-Reference: ruby-amqp/amq-protocol#72

### CODE REVIEW

**Comment by michaelklishin** [comment]
It does handle connection timeout, search for `set_pending_connect_timeout`.

**Comment by michaelklishin** [comment]
OK, at least one of the changes comes down to the fact that my `amq-protocol` dependency version is out of date :/

File: lib/amqp/session.rb [hunk 1]
```diff
@@ -157,7 +158,7 @@ def initialize(*args, &block)
 
       # EventMachine::Connection's and Adapter's constructors arity
       # make it easier to use *args. MK.
-      @settings                           = Settings.configure(args.first)
+      @settings                           = AMQ::Settings.configure(args.first)
```
Comment (michaelklishin) [inline lib/amqp/session.rb]: This completely ignored `other_options` propagated by `AMQP.connect`, which breaks error handling and makes at least one test fail.

**Comment by Tensho** [comment]
My bad, I forgot to open PR for `amq-protocol` – ruby-amqp/amq-protocol#72.

File: lib/amqp/session.rb [hunk 2]
```diff
@@ -436,8 +436,9 @@ def self.authentication_failure_exception_class
     # @option settings [Fixnum] :frame_max (131072) Maximum frame size to use. If broker cannot support frames this large, broker's maximum value will be used instead.
     #
     # @param [Hash] settings
-    def self.connect(settings = {}, &block)
-      @settings = Settings.configure(settings)
+    # def self.connect(settings = {}, &block)
+    def self.connect(connection_string_or_opts = ENV['RABBITMQ_URL'], other_options = {}, &block)
+      @settings = AMQ::Settings.configure(connection_string_or_opts)
```
Comment (michaelklishin) [inline lib/amqp/session.rb]: Same as above, the `other_options` argument is ignored.

File: lib/amqp/session.rb [hunk 3]
```diff
@@ -104,7 +105,7 @@ class Session < EM::Connection
     class << self
       # Settings
       def settings
-        @settings ||= AMQP::Settings.default
+        @settings ||= AMQ::Settings.default
```
Comment (Tensho) [inline lib/amqp/session.rb]: Do we need `AMQP.settings` method that fetches only `:logging` key? I checked the repo and didn't find any usage of it, but as far as it has comment about **global** behaviour, I thought one of underlying gems could rely on this parameter, from example `eventmachine`.

**Comment by michaelklishin** [comment]
I corrected a couple of things and merged manually into master.

**Comment by Tensho** [comment]
@michaelklishin I see, you are super streamline ^_^
I've added `connection_timeout` handling in 57ead6b05333c66f7843570452a44a6826f2dcc2, could you merge it also?

**Comment by michaelklishin** [comment]
Yup, cherry-picking as we speak.

**Comment by Tensho** [comment]
I've got the great experience, thank you for the support 🙇 
Happy Christmas and New Year, Michael 🎄

**Comment by michaelklishin** [comment]
@Tensho all done. Thanks for the greetings, likewise!

**Comment by ygnessin** [comment]
FYI this set of changes caused a breaking change in our app when we upgraded from 1.5.0 to 1.8.0. We had been using `AMQP::Settings.parse_connection_url`, which is removed in this PR. Because this isn't called out in the changelog, this error actually took me a long time to track down, and I figured out I needed to replace it with `AQM::Settings.parse_amqp_url` - so hopefully this comment will help others with the same issue.

For the future, I think it would be a good idea if breaking changes like this were saved for major version updates, e.g. 2.0, rather than minor version updates which are typically kept solely for adding features, not backwards-incompatible changes - if you follow semantic versioning practices. That could have avoided this situation, when a seemingly "minor" version update would have caused an outage on our production site, if I hadn't caught it early. At the very least, it would be nice if changes like that could be called out explicitly in the Change Log. Alternatively, you could have kept `AMQP::Settings` and its methods around with a deprecation warning, and had them delegate functionality to their new versions.

Nevertheless, I'd like to thank you for maintaining this gem and keeping it up to date. I hope you'll take my suggestions into consideration. Thanks for the great library!

**Pull request closed without merging**